src/execution/window_executor.cpp

```diff
@@ -0,0 +1,1280 @@
+	}
+
+	vector<LogicalType> types;
+	for (idx_t expr_idx = 0; expr_idx < exprs.size(); ++expr_idx) {
+		types.push_back(exprs[expr_idx]->return_type);
+		executor.AddExpression(*exprs[expr_idx]);
+	}
+
+	if (!types.empty()) {
+		auto &allocator = executor.GetAllocator();
+		chunk.Initialize(allocator, types);
+	}
+}
+
+WindowExecutor::WindowExecutor(BoundWindowExpression &wexpr, ClientContext &context, const idx_t payload_count,
+                               const ValidityMask &partition_mask, const ValidityMask &order_mask)
+    : wexpr(wexpr), context(context), payload_count(payload_count), partition_mask(partition_mask),
+      order_mask(order_mask), payload_collection(), payload_executor(context),
+      range((HasPrecedingRange(wexpr) || HasFollowingRange(wexpr)) ? wexpr.orders[0].expression.get() : nullptr,
+            context, payload_count) {
+	// TODO: child may be a scalar, don't need to materialize the whole collection then
+
+	// evaluate inner expressions of window functions, could be more complex
+	PrepareInputExpressions(wexpr.children, payload_executor, payload_chunk);
+
+	auto types = payload_chunk.GetTypes();
+	if (!types.empty()) {
+		payload_collection.Initialize(Allocator::Get(context), types);
+	}
+}
+
+unique_ptr<WindowExecutorState> WindowExecutor::GetExecutorState() const {
+	return make_uniq<WindowExecutorBoundsState>(wexpr, context, payload_count, partition_mask, order_mask);
+}
+
+//===--------------------------------------------------------------------===//
+// WindowAggregateExecutor
+//===--------------------------------------------------------------------===//
+bool WindowAggregateExecutor::IsConstantAggregate() {
+	if (!wexpr.aggregate) {
+		return false;
+	}
+
+	//	COUNT(*) is already handled efficiently by segment trees.
+	if (wexpr.children.empty()) {
+		return false;
+	}
+
+	/*
+	    The default framing option is RANGE UNBOUNDED PRECEDING, which
+	    is the same as RANGE BETWEEN UNBOUNDED PRECEDING AND CURRENT
+	    ROW; it sets the frame to be all rows from the partition start
+	    up through the current row's last peer (a row that the window's
+	    ORDER BY clause considers equivalent to the current row; all
+	    rows are peers if there is no ORDER BY). In general, UNBOUNDED
+	    PRECEDING means that the frame starts with the first row of the
+	    partition, and similarly UNBOUNDED FOLLOWING means that the
+	    frame ends with the last row of the partition, regardless of
+	    RANGE, ROWS or GROUPS mode. In ROWS mode, CURRENT ROW means that
+	    the frame starts or ends with the current row; but in RANGE or
+	    GROUPS mode it means that the frame starts or ends with the
+	    current row's first or last peer in the ORDER BY ordering. The
+	    offset PRECEDING and offset FOLLOWING options vary in meaning
+	    depending on the frame mode.
+	*/
+	switch (wexpr.start) {
+	case WindowBoundary::UNBOUNDED_PRECEDING:
+		break;
+	case WindowBoundary::CURRENT_ROW_RANGE:
+		if (!wexpr.orders.empty()) {
```

Comment:
This appears to be uncovered

Reply:
There are some conditions in here that can't actually happen, but I'll put them in the net PR to avoid changing code.